### PR TITLE
Lock decay angles tracks to common vertex.

### PR DESCRIPTION
**Pull request by Joseph-Garvey**
Edit (@palvarezc ):
The shape for the decay angles calculation is composed of three `Line` shapes. We want to maintain the decay vertex of the particle, common between the three lines. The only way that I managed to get something close is:

- Wait for an `layer.events.data` event
- If the common vertex has been moved for any of the shapes, move the corresponding vertex for the other shapes

The update of the other lines only happens once the user releases the vertex being moved, as that is when the event  of modification of the layer data is emitted. Something closer to a `mouse_drag` would be better, but could not find a way to implement it. I also wonder if in the future, we could just define a custom shape.

Also in this PR:
- Update angle calculation to give a signed value of the angle in radians.

### CODE REVIEW

File: src/cavendish_particle_tracks/_decay_angles_dialog.py [hunk 1]
```diff
@@ -17,27 +20,22 @@
...
-        bss.clicked.connect(self._on_click_calculate)
-
-        bap = QPushButton("Save to table")
-        bap.clicked.connect(self._on_click_save_to_table)
-
+        # buttons
+        btn_calculate = QPushButton("Calculate")
+        btn_calculate.clicked.connect(self._on_click_calculate)
+        btn_save = QPushButton("Save to table")
+        btn_save.clicked.connect(self._on_click_save_to_table)
```
Comment (samcunliffe) [inline src/cavendish_particle_tracks/_decay_angles_dialog.py:38]: Changes are off-topic from the PR title.

File: src/cavendish_particle_tracks/_decay_angles_dialog.py [hunk 2]
```diff
@@ -55,30 +53,83 @@ def __init__(self, parent=None):
...
-
+        self.layout().addWidget(btn_calculate, 5, 0, 1, 3)
         self.layout().addWidget(
             QLabel("Opening angles"),
-            5,
+            6,
             0,
             1,
             3,
         )
```
Comment (samcunliffe) [inline src/cavendish_particle_tracks/_decay_angles_dialog.py:63]: Good old sensible formatting 🫠 .

If you delete the final comma and run `pre-commit` will it put it all on one line now?

File: src/cavendish_particle_tracks/_decay_angles_dialog.py [hunk 3]
```diff
@@ -55,30 +53,83 @@ def __init__(self, parent=None):
...
             [QLabel("ϕ_proton [rad]"), QLabel("ϕ_pion [rad]")] + self.textboxes_phi
         ):
-            self.layout().addWidget(widget, i % 2 + 6, i // 2 + 1)
-        self.layout().addWidget(bap, 8, 0, 1, 3)
-        self.layout().addWidget(self.buttonBox, 10, 0, 1, 3)
-
+            self.layout().addWidget(widget, i % 2 + 7, i // 2 + 1)
+        self.layout().addWidget(btn_save, 9, 0, 1, 3)
+        self.layout().addWidget(self.buttonBox, 11, 0, 1, 3)
+        # endregion
```
Comment (samcunliffe) [inline src/cavendish_particle_tracks/_decay_angles_dialog.py:70]:
```suggestion

```
Again, the comment isn't clarifying anything.

File: src/cavendish_particle_tracks/_decay_angles_dialog.py [hunk 4]
```diff
@@ -55,30 +53,83 @@ def __init__(self, parent=None):
...
         )
         for i, widget in enumerate(
             [QLabel("ϕ_proton [rad]"), QLabel("ϕ_pion [rad]")] + self.textboxes_phi
         ):
-            self.layout().addWidget(widget, i % 2 + 6, i // 2 + 1)
-        self.layout().addWidget(bap, 8, 0, 1, 3)
-        self.layout().addWidget(self.buttonBox, 10, 0, 1, 3)
-
+            self.layout().addWidget(widget, i % 2 + 7, i // 2 + 1)
+        self.layout().addWidget(btn_save, 9, 0, 1, 3)
```
Comment (samcunliffe) [inline src/cavendish_particle_tracks/_decay_angles_dialog.py:68]: This looks like minor UI changes that we probably don't want conflated with this feature.

File: src/cavendish_particle_tracks/_decay_angles_dialog.py [hunk 5]
```diff
@@ -55,30 +53,83 @@ def __init__(self, parent=None):
...
+        #     if selected_shape==0 and (data[0][0] != data[1][0]).all():
+        #         # # This is what we want but needs private methods
+        #         # vertices = self.cal_layer._data_view.shapes[1].data
+        #         # vertices[0] = data[0][0]
+        #         # self.cal_layer._data_view.edit(1, vertices, new_type=None)
+        #         # self.cal_layer.refresh()
+        #         data[1][0][0] = data[0][0][0]
+        #         data[1][0][1] = data[0][0][1]
+        #         self.cal_layer.refresh()
+        # self.cal_layer.events.set_data.connect(enforce_points_coincident_v2)
```
Comment (samcunliffe) [inline src/cavendish_particle_tracks/_decay_angles_dialog.py:91]:
```suggestion
```
Don't want commented-out code in production.

File: src/cavendish_particle_tracks/_decay_angles_dialog.py [hunk 6]
```diff
@@ -55,30 +53,83 @@ def __init__(self, parent=None):
...
+                    if (i not in shapes_modified) and (
+                        self.cal_layer.data[i][0]
+                        != self.cal_layer.data[shapes_modified[0]][0]
+                    ).any():
+                        data[i][0] = self.cal_layer.data[shapes_modified[0]][0]
+                        self.cal_layer.data = data
+            else:
+                # nothing moved, or more shapes have been added
+                # do nothing
+                return
```
Comment (samcunliffe) [inline src/cavendish_particle_tracks/_decay_angles_dialog.py:131]: Flip the logic to _guard_ the active code. Then you can unindent (#90CharactersPerLine).

```py
if event_action is not "changed":
    return  # Nothing has been moved, so nothing to do

shapes_modified = event.data_indices
data = self.cal_layer.data

...
```

File: src/cavendish_particle_tracks/_decay_angles_dialog.py [hunk 7]
```diff
@@ -116,7 +168,7 @@ def _setup_decayangles_layer(self):
 
         text = {
             "string": ["Λ", "p", "π"],
-            "size": 14,
+            "size": 20,
```
Comment (samcunliffe) [inline src/cavendish_particle_tracks/_decay_angles_dialog.py:171]: Again, this looks like scope creep. Is the text currently too small?

File: src/cavendish_particle_tracks/_decay_angles_dialog.py [hunk 8]
```diff
@@ -135,11 +187,19 @@ def _setup_decayangles_layer(self):
...
         """When 'Calculate' button is clicked, calculate opening angles and populate table"""
 
+        # The Lambda travels towards the decay vertex, line needs to be reversed
+        lambda_line = self.cal_layer.data[0][::-1]
+        proton_line = self.cal_layer.data[1]
+        pion_line = self.cal_layer.data[2]
+
+        # Save the lines
+        self.alines = [lambda_line, proton_line, pion_line]
+
```
Comment (samcunliffe) [inline src/cavendish_particle_tracks/_decay_angles_dialog.py:197]: Why are these changes included?

File: tests/test_decay_angles_dialog.py [hunk 9]
```diff
@@ -45,6 +46,71 @@ def test_open_and_close_decay_angles_dialog(cpt_widget, qtbot: QtBot, click_twic
...
+
+    cpt_widget.particle_decays_menu.setCurrentIndex(4)
+    dialog = cpt_widget._on_click_decay_angles()
+
+    # Move the decay vertex of the Lambda
+    data = dialog.cal_layer.data
+    old_value = data[0][0][0]
+    data[0][0][0] = old_value + 50
+    dialog.cal_layer.data = data  # For some reason, this doesn't trigger the change event
+    dialog.cal_layer.events.data(action="changed", data_indices=(0,))
```
Comment (samcunliffe) [inline tests/test_decay_angles_dialog.py:59]: `cal_layer` is what?

Comment (samcunliffe) [inline tests/test_decay_angles_dialog.py:59]:
```suggestion
    # Move the decay vertex of the Lambda
    old_value = dialog.cal_layer.data[0][0][0]
    dialog.cal_layer.data[0][0][0] = old_value + 50
    # For some reason, this doesn't trigger the change event
    dialog.cal_layer.events.data(action="changed", data_indices=(0,))
```

Can remove some of the intermediary variable assignment... but in either case, do those first two lines trigger the change event?

File: tests/test_decay_angles_dialog.py [hunk 10]
```diff
@@ -45,6 +46,71 @@ def test_open_and_close_decay_angles_dialog(cpt_widget, qtbot: QtBot, click_twic
...
+
+    # Check that the decay vertex has been updated
+    assert dialog.cal_layer.data[0][0][0] == old_value + 50
+    assert (dialog.cal_layer.data[1][0] == data[0][0]).all()
+    assert (dialog.cal_layer.data[2][0] == data[0][0]).all()
```
Comment (samcunliffe) [inline tests/test_decay_angles_dialog.py:64]:
```suggestion
    assert dialog.cal_layer.data[0][0][0] == old_value + 50, "The decay vertex has not been updated!"
    assert (dialog.cal_layer.data[1][0] == data[0][0]).all()
    assert (dialog.cal_layer.data[2][0] == data[0][0]).all()
```

Can you add explanatory strings to the other guys too? I don't get what condition they're testing.

File: tests/test_decay_angles_dialog.py [hunk 11]
```diff
@@ -45,6 +46,71 @@ def test_open_and_close_decay_angles_dialog(cpt_widget, qtbot: QtBot, click_twic
...
+
+    # Move two tracks simultaneously
+    data = dialog.cal_layer.data
+    old_value = data[0][0][0]
+    data[1][0][0] = old_value + 50
+    data[2][0][0] = old_value + 50
+    dialog.cal_layer.data = data  # For some reason, this doesn't trigger the change event
+    dialog.cal_layer.events.data(action="changed", data_indices=(1, 2))
```
Comment (samcunliffe) [inline tests/test_decay_angles_dialog.py:72]:
```suggestion
    # Move two tracks simultaneously
    old_value = dialog.cal_layer.data[0][0][0]
    dialog.cal_layer.data[1][0][0] = old_value + 50
    dialog.cal_layer.data[2][0][0] = old_value + 50
    # For some reason, this doesn't trigger the change event
    dialog.cal_layer.events.data(action="changed", data_indices=(1, 2))
```

File: tests/test_decay_angles_dialog.py [hunk 12]
```diff
@@ -45,6 +46,71 @@ def test_open_and_close_decay_angles_dialog(cpt_widget, qtbot: QtBot, click_twic
...
+        (
+            [[0, 0], [-1, 0]],
+            [[0, 0], [1, 1]],
+            [[0, 0], [1, -1]],
+            np.pi / 4,
+            -1 * np.pi / 4,
+        ),
+    ],
+)
+def test_calculate_decay_angles_ui(
```
Comment (samcunliffe) [inline tests/test_decay_angles_dialog.py:90]: Was this just untested before?

**Review by samcunliffe — CHANGES_REQUESTED** [review]
It seems to work 👍 


https://github.com/user-attachments/assets/0b4c88b5-f347-477e-9369-af2923af46a0


But playing with the "Decay Angles Tool" layer by, e.g. moving up or down in the stack and changing the visibility, I can generate a `TypeError` about subscripting a NoneType. (The very end of that video.) And I cannot generate it on the current `HEAD` of `main`.

<details><summary>Traceback</summary>

```
File /opt/homebrew/Caskroom/miniforge/base/envs/napari/lib/python3.12/site-packages/napari/utils/events/event.py:764, in EventEmitter.__call__(self=<napari.utils.events.event.EventEmitter object>, *args=(), **kwargs={})
    761     self._block_counter.update([cb])
    762     continue
--> 764 self._invoke_callback(cb, event if pass_event else None)
        event = <Event blocked=False handled=False native=None source=None sources=[] type='highlight'>
        self = <napari.utils.events.event.EventEmitter object at 0x146aab680>
        cb = <bound method VispyShapesLayer._on_highlight_change of <napari._vispy.layers.shapes.VispyShapesLayer object at 0x146426c90>>
        pass_event = False
    765 if event.blocked:
    766     break

File /opt/homebrew/Caskroom/miniforge/base/envs/napari/lib/python3.12/site-packages/napari/utils/events/event.py:802, in EventEmitter._invoke_callback(self=<napari.utils.events.event.EventEmitter object>, cb=<bound method VispyShapesLayer._on_highlight_cha...ri._vispy.layers.shapes.VispyShapesLayer object>>, event=None)
    800     self.disconnect(cb)
    801     return
--> 802 _handle_exception(
        self = <napari.utils.events.event.EventEmitter object at 0x146aab680>
        event = None
        cb = <bound method VispyShapesLayer._on_highlight_change of <napari._vispy.layers.shapes.VispyShapesLayer object at 0x146426c90>>
        (cb, event) = (<bound method VispyShapesLayer._on_highlight_change of <napari._vispy.layers.shapes.VispyShapesLayer object at 0x146426c90>>, None)
    803     self.ignore_callback_errors,
    804     self.print_callback_errors,
    805     self,
    806     cb_event=(cb, event),
    807 )

File /opt/homebrew/Caskroom/miniforge/base/envs/napari/lib/python3.12/site-packages/napari/utils/events/event.py:791, in EventEmitter._invoke_callback(self=<napari.utils.events.event.EventEmitter object>, cb=<bound method VispyShapesLayer._on_highlight_cha...ri._vispy.layers.shapes.VispyShapesLayer object>>, event=None)
    789         cb(event)
    790     else:
--> 791         cb()
        cb = <bound method VispyShapesLayer._on_highlight_change of <napari._vispy.layers.shapes.VispyShapesLayer object at 0x146426c90>>
    792 except Exception as e:  # noqa: BLE001
    793     # dead Qt object with living python pointer. not importing Qt
    794     # here... but this error is consistent across backends
    795     if (
    796         isinstance(e, RuntimeError)
    797         and 'C++' in str(e)
    798         and str(e).endswith(('has been deleted', 'already deleted.'))
    799     ):

File /opt/homebrew/Caskroom/miniforge/base/envs/napari/lib/python3.12/site-packages/napari/_vispy/layers/shapes.py:92, in VispyShapesLayer._on_highlight_change(self=<napari._vispy.layers.shapes.VispyShapesLayer object>)
     78 self.node.shape_highlights.set_data(
     79     vertices=vertices,
     80     faces=faces,
     81     color=self.layer._highlight_color,
     82 )
     84 # Compute the location and properties of the vertices and box that
     85 # need to get rendered
     86 (
     87     vertices,
     88     face_color,
     89     edge_color,
     90     pos,
     91     _,
---> 92 ) = self.layer._compute_vertices_and_box()
        vertices = array([[0., 0.],
       [0., 0.],
       [0., 0.]])
        self.layer = <Shapes layer 'Decay Angles Tool' at 0x146a9f860>
        self = <napari._vispy.layers.shapes.VispyShapesLayer object at 0x146426c90>
     94 width = settings.appearance.highlight.highlight_thickness
     96 if vertices is None or len(vertices) == 0:

File /opt/homebrew/Caskroom/miniforge/base/envs/napari/lib/python3.12/site-packages/napari/layers/shapes/shapes.py:2542, in Shapes._compute_vertices_and_box(self=<Shapes layer 'Decay Angles Tool'>)
   2539 if self._mode == Mode.ADD_PATH:
   2540     vertices = vertices[:-1]
-> 2542 if self._value[0] is None or self._value[1] is None:
        self = <Shapes layer 'Decay Angles Tool' at 0x146a9f860>
        self._value = None
   2543     face_color = 'white'
   2544 else:

TypeError: 'NoneType' object is not subscriptable
```

</details>


I think we should remove all off-topic changes from this PR before debugging further. Probably also add (or tweak) the tests to toggle visibility of the layer (and move its position in the stack?)

File: src/cavendish_particle_tracks/_decay_angles_dialog.py (hunk 4 above)
Comment (palvarezc) [reply src/cavendish_particle_tracks/_decay_angles_dialog.py:68]: There are some changes needed in the UI, as I needed to add a new combobox for the choice of `origin` vs `decay` vertex. Therefore, I'd go ahead and update the button names.

File: src/cavendish_particle_tracks/_decay_angles_dialog.py [hunk 13]
```diff
@@ -17,27 +20,22 @@
 class DecayAnglesDialog(QDialog):
     def __init__(self, parent=None):
         super().__init__(parent)
-
         self.parent = parent
-
+        # region UI Setup
```
[the remark this answers is not in the document]
Comment (palvarezc) [reply src/cavendish_particle_tracks/_decay_angles_dialog.py:24]: OK, will change these.

File: src/cavendish_particle_tracks/_decay_angles_dialog.py (hunk 1 above)
Comment (palvarezc) [reply src/cavendish_particle_tracks/_decay_angles_dialog.py:38]: I'd keep them as some modifications to the UI _are_ needed.

File: src/cavendish_particle_tracks/_decay_angles_dialog.py (hunk 5 above)
Comment (palvarezc) [reply src/cavendish_particle_tracks/_decay_angles_dialog.py:91]: OK, I'll open another PR.

File: src/cavendish_particle_tracks/_decay_angles_dialog.py (hunk 7 above)
Comment (palvarezc) [reply src/cavendish_particle_tracks/_decay_angles_dialog.py:171]: Yes.

File: src/cavendish_particle_tracks/_decay_angles_dialog.py (hunk 8 above)
Comment (palvarezc) [reply src/cavendish_particle_tracks/_decay_angles_dialog.py:197]: Because the function that connects the decay vertex of the Λ and the origin vertices of the proton and π is much simpler if the vertex index is the same for all the `Line` objects. Hence, the Λ's direction needs to be reverted for the angle calculation.

File: tests/test_decay_angles_dialog.py (hunk 12 above)
Comment (palvarezc) [reply tests/test_decay_angles_dialog.py:87]: Yup...

File: src/cavendish_particle_tracks/_decay_angles_dialog.py (hunk 8 above)
Comment (samcunliffe) [reply src/cavendish_particle_tracks/_decay_angles_dialog.py:197]: Ah. Ok, fair enough.

File: tests/test_decay_angles_dialog.py (hunk 9 above)
Comment (palvarezc) [reply tests/test_decay_angles_dialog.py]: The `Decay Angles Tool` layer.

**Comment by palvarezc** [comment]
> But playing with the "Decay Angles Tool" layer by, e.g. moving up or down in the stack and changing the visibility, I can generate a `TypeError` about subscripting a NoneType. (The very end of that video.) And I cannot generate it on the current `HEAD` of `main`.

I added a `not None` check for the event, this seems to fix it....?

**Comment by samcunliffe** [comment]
> > But playing with the "Decay Angles Tool" layer by, e.g. moving up or down in the stack and changing the visibility, I can generate a `TypeError` about subscripting a NoneType. (The very end of that video.) And I cannot generate it on the current `HEAD` of `main`.
> 
> I added a `not None` check for the event, this seems to fix it....?

Nice!

But....
Screenshot 2024-09-30 at 15 44 33

It'd be good to make sure we can trigger that in a test reliably...

File: tests/test_decay_angles_dialog.py (hunk 9 above)
Comment (samcunliffe) [reply tests/test_decay_angles_dialog.py]: What does `cal` mean? Can I get a "Rename symbol" to ... `dialog.decay_angles_layer` or something?

**Review by samcunliffe — APPROVED** [review]
